plugin/gcp_account_resources.go

```diff
@@ -0,0 +1,135 @@
+	"github.com/hashicorp/go-gcp-common/gcputil"
+	"github.com/hashicorp/vault-plugin-secrets-gcp/plugin/iamutil"
+	"github.com/hashicorp/vault-plugin-secrets-gcp/plugin/util"
+	"github.com/hashicorp/vault/sdk/helper/useragent"
+	"github.com/hashicorp/vault/sdk/logical"
+	"google.golang.org/api/iam/v1"
+)
+
+type (
+	// gcpAccountResources is a wrapper around the GCP resources Vault creates to generate credentials.
+	// This includes a Vault-managed GCP service account (required), IAM bindings, and/or key via TokenGenerator
+	// (for generating access tokens).
+	gcpAccountResources struct {
+		accountId gcputil.ServiceAccountId
+		bindings  ResourceBindings
+		tokenGen  *TokenGenerator
+	}
+
+	// ResourceBindings represent a map of GCP resource name to IAM roles to be bound on that resource.
+	ResourceBindings map[string]util.StringSet
+
+	// TokenGenerator wraps the service account key and params required to create access tokens.
+	TokenGenerator struct {
+		KeyName    string
+		B64KeyJSON string
+		Scopes     []string
+	}
+)
+
+func (rb ResourceBindings) asOutput() map[string][]string {
+	out := make(map[string][]string)
+	for k, v := range rb {
+		out[k] = v.ToSlice()
+	}
+	return out
+}
+
+func getStringHash(bindingsRaw string) string {
+	ssum := sha256.Sum256([]byte(bindingsRaw)[:])
+	return base64.StdEncoding.EncodeToString(ssum[:])
+}
+
+func (b *backend) createNewTokenGen(ctx context.Context, req *logical.Request, parent string, scopes []string) (*TokenGenerator, error) {
+	b.Logger().Debug("creating new TokenGenerator (service account key)", "account", parent, "scopes", scopes)
+
+	iamAdmin, err := b.IAMAdminClient(req.Storage)
+	if err != nil {
+		return nil, err
+	}
+
+	key, err := iamAdmin.Projects.ServiceAccounts.Keys.Create(
+		parent,
+		&iam.CreateServiceAccountKeyRequest{
+			PrivateKeyType: privateKeyTypeJson,
+		}).Do()
+	if err != nil {
+		return nil, err
+	}
+	return &TokenGenerator{
+		KeyName:    key.Name,
+		B64KeyJSON: key.PrivateKeyData,
+		Scopes:     scopes,
+	}, nil
+}
+
+func (b *backend) createIamBindings(ctx context.Context, req *logical.Request, saEmail string, binds ResourceBindings) error {
+	b.Logger().Debug("creating IAM bindings", "account_email", saEmail)
+	httpC, err := b.HTTPClient(req.Storage)
+	if err != nil {
+		return err
+	}
+	iamHandle := iamutil.GetIamHandle(httpC, useragent.String())
+
+	for resourceName, roles := range binds {
+		b.Logger().Debug("setting IAM binding", "resource", resourceName, "roles", roles)
+		resource, err := b.iamResources.Parse(resourceName)
+		if err != nil {
+			return err
+		}
+
+		b.Logger().Debug("getting IAM policy for resource name", "name", resourceName)
+		p, err := iamHandle.GetIamPolicy(ctx, resource)
+		if err != nil {
+			return nil
+		}
+
+		b.Logger().Debug("got IAM policy for resource name", "name", resourceName)
+		changed, newP := p.AddBindings(&iamutil.PolicyDelta{
+			Roles: roles,
+			Email: saEmail,
+		})
+		if !changed || newP == nil {
```

Comment:
Can `newP` ever be `nil`?

Reply:
yes/no (changed would be false, so would catch) - i don't see any harm in keeping the check though in case?

Reply:
It's OK as is. Mainly was curious.